rust/src/mysql/parser.rs

```diff
@@ -0,0 +1,3350 @@
+
+        Ok((i, MysqlResultBinarySetRow::Text(texts)))
+    }
+}
+
+fn parse_response_resultset(i: &[u8], n_cols: u64) -> IResult<&[u8], MysqlResponse> {
+    let (i, columns) = many_m_n(n_cols as usize, n_cols as usize, parse_column_definition)(i)?;
+    let (i, eof) = parse_eof_packet(i)?;
+    let (i, (rows, _)) = many_till(parse_resultset_row, |i| {
+        let (i, header) = parse_packet_header(i)?;
+        let (i, response_code) = verify(be_u8, |&x| x == 0xFE || x == 0xFF)(i)?;
+        match response_code {
+            // EOF
+            0xFE => {
+                let (i, _) = cond(header.pkt_len - 1 > 0, take(header.pkt_len - 1))(i)?;
+                Ok((
+                    i,
+                    MysqlResponse {
+                        item: MysqlResponsePacket::EOF,
+                    },
+                ))
+            }
+            // ERR
+            0xFF => parse_response_err(i, header.pkt_len - 1),
+            _ => {
+                let (i, _) = cond(header.pkt_len - 1 > 0, take(header.pkt_len - 1))(i)?;
+                Ok((
+                    i,
+                    MysqlResponse {
+                        item: MysqlResponsePacket::Unknown,
+                    },
+                ))
+            }
+        }
+    })(i)?;
+    Ok((
+        i,
+        MysqlResponse {
+            item: MysqlResponsePacket::ResultSet {
+                n_cols,
+                columns,
+                eof,
+                rows,
+            },
+        },
+    ))
+}
+
+fn parse_response_binary_resultset(i: &[u8], n_cols: u64) -> IResult<&[u8], MysqlResponse> {
+    let (i, columns) = many_m_n(n_cols as usize, n_cols as usize, parse_column_definition)(i)?;
+    let (i, eof) = parse_eof_packet(i)?;
+    let (i, (rows, _)) = many_till(parse_binary_resultset_row(columns), |i| {
+        // eof
+        let (i, header) = parse_packet_header(i)?;
+        let (i, response_code) = verify(be_u8, |&x| x == 0xFE || x == 0xFF)(i)?;
+        match response_code {
+            // EOF
+            0xFE => {
+                let (i, _) = cond(header.pkt_len - 1 > 0, take(header.pkt_len - 1))(i)?;
+                Ok((
+                    i,
+                    MysqlResponse {
+                        item: MysqlResponsePacket::EOF,
+                    },
+                ))
+            }
+            // ERR
+            0xFF => parse_response_err(i, header.pkt_len - 1),
+            _ => {
+                let (i, _) = cond(header.pkt_len - 1 > 0, take(header.pkt_len - 1))(i)?;
+                Ok((
+                    i,
+                    MysqlResponse {
+                        item: MysqlResponsePacket::Unknown,
+                    },
+                ))
+            }
+        }
+    })(i)?;
+    Ok((
+        i,
+        MysqlResponse {
+            item: MysqlResponsePacket::BinaryResultSet { n_cols, eof, rows },
+        },
+    ))
+}
+
+fn parse_response_ok(i: &[u8], length: u32) -> IResult<&[u8], MysqlResponse> {
```

Comment:
As above, some fields are flags dependent and some condition should be check before parsing them.
Reference: https://dev.mysql.com/doc/dev/mysql-server/latest/page_protocol_basic_ok_packet.html

Reply:
When i parse ok response,  i just take the fields what I needed, as mentioned before, hen i parse handshake response, i have checked capabilities should contain CLIENT_PROTOCOL_41.

Reply:
Fixed, please check again.